source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -476,41 +476,112 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
+    // ZooKeeper filter buffer contains partial packet data from the previous network filter buffer.
+    // Prepending ZooKeeper filter buffer to the current network filter buffer can help to generate
+    // full packets.
+    data.prepend(zk_filter_buffer);
+    decodeAndBufferHelper(data, zk_filter_buffer, dtype);
+    // Drain the prepended ZooKeeper filter buffer.
+    data.drain(zk_filter_buffer_len);
+  } else {
+    zk_filter_buffer.add(data);
+  }
+  return Network::FilterStatus::Continue;
+}
+
+void DecoderImpl::decodeAndBufferHelper(Buffer::Instance& data, Buffer::OwnedImpl& zk_filter_buffer,
+                                        DecodeType dtype) {
+  ASSERT(dtype == DecodeType::READ || dtype == DecodeType::WRITE);
+
+  uint32_t data_len = data.length();
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
   uint64_t offset = 0;
-  int32_t len = 0;
+  uint32_t len = 0;
+  // Boolean to check whether there is at least one full packet in the network filter buffer (to
+  // which the ZooKeeper filter buffer is prepended).
+  bool has_full_packets = false;
 
-  while (offset < data.length()) {
+  while (offset < data_len) {
     try {
       // Peek packet length.
       len = helper_.peekInt32(data, offset);
-      ensureMinLength(len, INT_LENGTH + XID_LENGTH);
+      ensureMinLength(len, dtype == DecodeType::READ ? XID_LENGTH + INT_LENGTH
+                                                     : XID_LENGTH + ZXID_LENGTH + INT_LENGTH);
       ensureMaxLength(len);
       offset += len;
+      if (offset <= data_len) {
+        has_full_packets = true;
+      }
     } catch (const EnvoyException& e) {
       ENVOY_LOG(debug, "zookeeper_proxy: decoding exception {}", e.what());
       callbacks_.onDecodeError();
-      return Network::FilterStatus::Continue;
+      return;
     }
   }
 
-  if (offset == data.length()) {
+  if (offset == data_len) {
     decode(data, dtype);
-    return Network::FilterStatus::Continue;
-  }
 
-  ENVOY_LOG(trace,
-            "zookeeper_proxy: waiting for entire packets in the buffer, current buffer "
-            "is {} bytes",
-            data.length());
-  return Network::FilterStatus::StopIteration;
+    if (zk_filter_buffer_len > 0) {
+      // After prepending the ZooKeeper filter buffer to the network filter buffer, the network
+      // filter buffer here includes one or more full packets without any partial data, the decoder
+      // is able to decode the entire network filter buffer. So the we need to drain the ZooKeeper
+      // filter buffer here, since its data has been decoded already. Otherwise, the ZooKeeper
+      // filter will prepend the ZooKeeper filter buffer to the network filter buffer again next
+      // time, which is not what we want.
+      zk_filter_buffer.drain(zk_filter_buffer_len);
+    }
+  } else if (offset > data_len) {
+    std::string temp_data;
+
+    if (has_full_packets) {
+      offset -= INT_LENGTH + len;
+      // Decode full packets.
```

Comment:
add comment 
`// TODO(Winbobob): use BufferFragment to avoid copying the full packets.`